### PR TITLE
Implement groupby `MERGE_LISTS` and `MERGE_SETS` aggregates

### LINKED ISSUE #7839: [FEA] Support `MERGE_LISTS` and `MERGE_SETS` for `groupby::aggregate`
**Is your feature request related to a problem? Please describe.**
To perform collect aggregations in Spark, (like other aggregation functions) it needs to go through two phases. In phase one, we perform collect aggregation within each node locally, which can be implemented by cuDF group_by with AggOp `collect_list` or `collect_set`. In phase two, we need to merge partial aggregation results from each node. In terms of collect aggregations, we concatenate multiple lists/sets into one. For instance,
```
Node A has a partial result of collect_list in group_by: [('A', [1, 2, 3]), ('B', [10, 20])]
Node B has a result: [('B', [30, 40]), ('C', [100, 200])]
Node C has a result: [('A', [4]), ('C', [300, 400])]

For phase two, we have a combined input table: [('A', [1, 2, 3]), ('B', [10, 20]), ('B', [30, 40]), ('C', [100, 200]), ('A', [4]), ('C', [300, 400])]
Then, we concatenate lists sharing same group key and we suppose to get the result:  [('A', [1, 2, 3, 4]), ('B', [10, 20, 30, 40]), ('C', [100, 200, 300, 400])]
```
I think we need two additional AggOps (`concatenate_list` and `concatenate_set`) for current feature.

### PR DESCRIPTION
**Pull request by ttnghia**
Groupby aggregations can be performed for distributed computing by the following approach:
 * Divide the dataset into batches
 * Run separate (distributed) aggregations over those batches on the distributed nodes
 * Merge the results of the step above into one final result by calling `groupby::aggregate` a final time on the master node

This PR supports merging operations for the lists resulted from distributed aggregate `collect_list` and `collect_set`.

Closes #7839.

### CODE REVIEW

File: cpp/tests/groupby/merge_lists_tests.cpp
Comment (ttnghia) [inline cpp/tests/groupby/merge_lists_tests.cpp:51]: This is a simple usage example of the new aggregate.

**Comment by sperlingxx** [comment]
LGTM, so far.

File: cpp/include/cudf/aggregation.hpp
Comment (jrhemstad) [inline cpp/include/cudf/aggregation.hpp:291]: I don't understand why this can't handle nullable input list columns.

File: cpp/include/cudf/aggregation.hpp
Comment (jrhemstad) [inline cpp/include/cudf/aggregation.hpp:315]: stale comment?

File: cpp/src/groupby/sort/group_merge_lists.cu
Comment (jrhemstad) [inline cpp/src/groupby/sort/group_merge_lists.cu:58]: This is just a gather?

File: cpp/include/cudf/aggregation.hpp
Comment (ttnghia) [reply cpp/include/cudf/aggregation.hpp:291]: Technically it can. But it is designed only to merge the results of collect_list/collect_set, which are not nullable. Thus, if its input is nullable then obviously it is operating on a wrong input data.

File: cpp/include/cudf/aggregation.hpp
Comment (ttnghia) [reply cpp/include/cudf/aggregation.hpp:315]: Wrong comment instead. Thanks for pointing this out.

File: cpp/src/groupby/sort/group_merge_lists.cu
Comment (ttnghia) [reply cpp/src/groupby/sort/group_merge_lists.cu:58]: Yes, similar to gather. More specifically, this is an upgrading of the grandchild column (of the input lists column) to be the child column of the output lists column.

File: cpp/include/cudf/aggregation.hpp
Comment (ttnghia) [reply cpp/include/cudf/aggregation.hpp:291]: I'll add more comments to the doxygen to clarify this.

File: cpp/src/groupby/sort/group_merge_lists.cu
Comment (jrhemstad) [reply cpp/src/groupby/sort/group_merge_lists.cu:58]: My point was that it'd be more descriptive to use a `thrust::gather` instead of a `thrust::transform`.

File: cpp/include/cudf/aggregation.hpp
Comment (jrhemstad) [reply cpp/include/cudf/aggregation.hpp:291]: Why can't `collect_list` be nullable? If a group has a nullable element in it, won't the resulting list have a null element?

File: cpp/src/groupby/sort/group_merge_lists.cu
Comment (ttnghia) [reply cpp/src/groupby/sort/group_merge_lists.cu:58]: Ah right! I'll do that.

File: cpp/include/cudf/aggregation.hpp
Comment (ttnghia) [reply cpp/include/cudf/aggregation.hpp:291]: No, the output will be lists collecting the elements. If we have nulls and:
* If the null handling parameter is `INCLUDED`: the output will be lists like `[null, ....null, ....]`
* If the null handling parameter is `EXCLUDED`: the output will be empty lists.

File: cpp/include/cudf/aggregation.hpp
Comment (sperlingxx) [inline cpp/include/cudf/aggregation.hpp:301]: If this aggregation requires non-nullable input values, why we still have the `null_equality` argument?

Comment (ttnghia) [reply cpp/include/cudf/aggregation.hpp:301]: It's explained below:
```
 * Similar to `MERGE_LISTS`, this aggregation does not need the `null_policy` parameter and requires * the input lists column to be non-nullable (its child column containing list entries is not * subject to this requirement). * * Since duplicate list entries will be removed, the parameters `null_equality` and `nan_equality` * are needed for calling to `lists::drop_list_duplicates`.
 ```

**Review by jrhemstad — APPROVED** [review]

File: cpp/include/cudf/aggregation.hpp
Comment (harrism) [inline cpp/include/cudf/aggregation.hpp:322]: These are all implementation details, and should not be in the public documentation. Instead, please clearly document the semantics and parameters visible to the caller.

The implementation details comments can be moved to the implementation, where they will be more helpful.

File: cpp/src/groupby/sort/group_reductions.hpp
Comment (harrism) [inline cpp/src/groupby/sort/group_reductions.hpp:373]:
```suggestion
 * @brief Internal API to merge grouped lists into one list.
```

File: cpp/tests/groupby/merge_sets_tests.cpp
Comment (harrism) [inline cpp/tests/groupby/merge_sets_tests.cpp:62]: These are repeated from the other file. Factor out?

**Review by harrism — CHANGES_REQUESTED** [review]
Good work, tests look very comprehensive. A couple of docs suggestions.

File: cpp/tests/groupby/merge_sets_tests.cpp
Comment (ttnghia) [reply cpp/tests/groupby/merge_sets_tests.cpp:62]: The original names are so long to use (such as `cudf::test::iterator_no_null()`). I want to have shorter names instead, but it's more difficult to change the existing APIs than aliasing them into local functions with much cleaner and shorter names.

File: cpp/include/cudf/aggregation.hpp
Comment (ttnghia) [reply cpp/include/cudf/aggregation.hpp]: Thanks. I have moved the detailed doxygen to detail implementation (https://github.com/rapidsai/cudf/pull/8436/files#diff-a943ede169c709882f6825c90823e342e974bbb95ef1a5ebe1bea7f083caf811R407).

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

File: cpp/tests/groupby/merge_sets_tests.cpp
Comment (harrism) [reply cpp/tests/groupby/merge_sets_tests.cpp:62]: That's fine, but two .cpp files in the same folder have the same functions. These two .cpp files end up in the same test binary, I believe, so all you need to do is only declare them in both of the .cpp files, but define them only once.

Comment (ttnghia) [reply cpp/tests/groupby/merge_sets_tests.cpp:62]: Thanks Mark. Unfortunately, these functions return `thrust::transform_iterator` thus we can't declare them multiple times. Thanks for your suggestion, I have another idea that can help to improve the entire testing suite better (which also benefit our future development). I just pushed a PR for it here: https://github.com/rapidsai/cudf/pull/8540

Comment (ttnghia) [reply cpp/tests/groupby/merge_sets_tests.cpp:62]: The PR https://github.com/rapidsai/cudf/pull/8540 blocks this one. After it is merged, I will refactor this one to use the new implementation.

Comment (harrism) [reply cpp/tests/groupby/merge_sets_tests.cpp:62]: > Thanks Mark. Unfortunately, these functions return thrust::transform_iterator thus we can't declare them multiple times.

Umm, you can make a header.

Comment (ttnghia) [reply cpp/tests/groupby/merge_sets_tests.cpp]: Done. These repeated functions are now gone :)

**Review by harrism — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**
